src/app/upgrade.cpp

```diff
@@ -53,19 +53,12 @@ namespace
             if (!newData.isEmpty() || oldData.isEmpty())
                 return;
 
-            QFile file(savePath);
-            if (!file.open(QIODevice::WriteOnly))
+            QSaveFile file(savePath);
+            if (!file.open(QIODevice::WriteOnly) || (file.write(oldData) != oldData.size()) || !file.commit())
             {
                 LogMsg(errorMsgFormat.arg(savePath, file.errorString()) , Log::WARNING);
                 return;
             }
-            if (file.write(oldData) != oldData.size())
-            {
-                file.close();
-                Utils::Fs::forceRemove(savePath);
```

Comment:
I would still keep this logic (i.e. try to delete file in case of error).

Reply:
>I would still keep this logic (i.e. try to delete file in case of error).

The logic is still there, from Qt doc:
>If commit() was not called and the QSaveFile instance is destroyed, the temporary file is discarded.

Reply:
>The logic is still there, from Qt doc:
>
>>If commit() was not called and the QSaveFile instance is destroyed, the temporary file is discarded.

I care about the case when data is successfully written in temporary file but QSaveFile::commit() fails to properly create destination file.

Reply:
>but QSaveFile::commit() fails to properly create destination file.

How would you detect it? `QSaveFile::commit()` returned `bool` isn't what you want in this case.

Reply:
>`QSaveFile::commit()` returned `bool` isn't what you want in this case.

Why?

Reply:
> 
> 
> > `QSaveFile::commit()` returned `bool` isn't what you want in this case.
> 
> Why?

https://doc.qt.io/qt-5/qsavefile.html#commit
>Commits the changes to disk, if all previous writes were successful.
>If an error happened during writing, deletes the temporary file and returns false. Otherwise, renames it to the final fileName and returns true on success. Finally, closes the device.

My interpretation is: it only detects previous writes and assumes the rename operation will be always successful.

Reply:
But anyway, `QSaveFile::commit()` will delete the temp file and it addresses [your concern](https://github.com/qbittorrent/qBittorrent/pull/15419#discussion_r702545994), isn't?

Reply:
>Otherwise, renames it to the final fileName and returns true on success. Finally, closes the device.

My interpretation is: it only returns `true` if rename operation is successful.
>it will delete the temp file and that addresses your concern, isn't?

No. My concern isn't about temp file but final/destination one.
Note: in some cases we have issue of broken `.fastresume` files which leads me to the idea that an error may still occur when renaming (otherwise the target files would not be affected at all).

Reply:
>No. My concern isn't about temp file but final/destination one.

It isn't clear to me what you exactly want. When `QSaveFile::commit()` returns `true`, it means the file is renamed successfully, then everything is fine, isn't? If you still think the file isn't renamed or there could still be error elsewhere, then as I said before `QSaveFile::commit()` won't be enough for you. You'll need other measures to verify it.

Reply:
>When `QSaveFile::commit()` returns `true`, it means the file is renamed successfully, then everything is fine, isn't? 

Yes.
I talk about the case when `QSaveFile::commit()` returns `false`.

Reply:
>I talk about the case when QSaveFile::commit() returns false.

Sounds like there are no more issues here then.

Reply:
>Sounds like there are no more issues here then.

The possible issue I talk about is when temporary file written successfully but there is error when renaming it to final file so we have broken destination file in the end. I can't imagine the exact conditions under which this can happen. I only base my judgments on existing reports of similar problems with `.fastresume` files.

Reply:
>but there is error when renaming it to final file so we have broken destination file in the end.

maybe it is: https://bugreports.qt.io/browse/QTBUG-75077
If that is the case I would defer it to another PR. This PR moves all QSaveFile usage to a single place which is enough for now. It would be easy to update it (the single place) in the future.

Reply:
Ok. It seems that contrary to my assumption, the decision to migrate these settings does not depend on the existence of this file, so this should not be a problem.